### PR TITLE
feat(clients/go): add evaluate decision command

### LINKED ISSUE #11049: Support standalone Decision evaluation in the Go client
**Is your feature request related to a problem? Please describe.**
As a user, I can trigger a standalone Decision evaluation through the Go client.

**Describe the solution you'd like**
- [ ] Add an "evaluateDecision" command in the Go client.

**Describe alternatives you've considered**
/

**Additional context**
/

**Blocked by**
* https://github.com/camunda/zeebe/issues/11047

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Support triggering standalone decision evaluation through the `evaluateDecision` command in the Go client.

## Related issues

closes #11049 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: clients/go/pkg/commands/evaluate_decision.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+import (
+	"context"
+	"fmt"
+	"github.com/camunda/zeebe/clients/go/v8/internal/utils"
+	"github.com/camunda/zeebe/clients/go/v8/pkg/pb"
+)
+
+type DispatchEvaluateDecisionCommand interface {
+	Send(context.Context) (*pb.EvaluateDecisionResponse, error)
```
Comment (remcowesterhoud) [inline clients/go/pkg/commands/evaluate_decision.go:25]: ❓  Why do we need this interface? Can't we just put the send on the step2 interface?

File: clients/go/pkg/commands/evaluate_decision_test.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+import (
+	"context"
+	"github.com/camunda/zeebe/clients/go/v8/internal/mock_pb"
+	"github.com/camunda/zeebe/clients/go/v8/internal/utils"
+	"github.com/camunda/zeebe/clients/go/v8/pkg/pb"
+	"github.com/golang/mock/gomock"
+	"testing"
+)
+
+func TestEvaluateDecisionCommand(t *testing.T) {
```
Comment (remcowesterhoud) [inline clients/go/pkg/commands/evaluate_decision_test.go:26]:
```suggestion
func TestEvaluateDecisionCommandByKey(t *testing.T) {
```

🔧 A bit nitpicky

File: clients/go/pkg/pb/gateway.pb.go [hunk 3]
```diff
@@ -121,7 +121,7 @@ func (x Partition_PartitionBrokerHealth) Number() protoreflect.EnumNumber {
 
 // Deprecated: Use Partition_PartitionBrokerHealth.Descriptor instead.
 func (Partition_PartitionBrokerHealth) EnumDescriptor() ([]byte, []int) {
-	return file_gateway_proto_rawDescGZIP(), []int{33, 1}
+	return file_gateway_proto_rawDescGZIP(), []int{39, 1}
```
Comment (remcowesterhoud) [inline clients/go/pkg/pb/gateway.pb.go:124]: ❓ What does this number do?

Comment (remcowesterhoud) [reply clients/go/pkg/pb/gateway.pb.go:124]: And all the other random number changes 😄

**Review by remcowesterhoud — APPROVED** [review]
Go syntax is horrible 😄 I found it somewhat hard to review, but comparing it to other command it looks to be in line with it. I do have some questions and a minor comment.

Have a look at them and feel free to merge 🚀

File: clients/go/pkg/pb/gateway.pb.go (hunk 3 above)
Comment (remcowesterhoud) [reply clients/go/pkg/pb/gateway.pb.go:124]: Is this partly auto-generated or something 🤔

File: engine/src/main/java/io/camunda/zeebe/engine/processing/dmn/EvaluateDecisionProcessor.java [hunk 4]
```diff
@@ -91,19 +91,14 @@ private Either<Failure, PersistedDecision> getDecision(final DecisionEvaluationR
...
-    if (decisionIdProvided) {
+    if (!decisionId.isEmpty()) {
       return decisionBehavior.findDecisionById(decisionId);
       // TODO: expand DecisionState API to find decisions by ID AND VERSION (#11230)
-    } else {
+    } else if (decisionKey > -1L) {
       return decisionBehavior.findDecisionByKey(decisionKey);
+    } else {
+      // if both ID and KEY are missing
+      return Either.left(new Failure(ERROR_MESSAGE_NO_IDENTIFIER_SPECIFIED));
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/dmn/EvaluateDecisionProcessor.java:101]: 🔧 It is a bit strange to me that we prioritise the decision id now. The key is more specific.

File: clients/go/pkg/pb/gateway.pb.go (hunk 3 above)
Comment (koevskinikola) [reply clients/go/pkg/pb/gateway.pb.go:124]: yes, it was auto-generated. I have no idea why it was changed.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/dmn/EvaluateDecisionProcessor.java (hunk 4 above)
Comment (koevskinikola) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/dmn/EvaluateDecisionProcessor.java:101]: It's the same for creating a process instance ([src](https://github.com/camunda/zeebe/blob/f549e5521664277517abb20f369f17edf47e2eb5/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceProcessor.java#L334-L345)). 

I suspect it's related to the issue I'm facing: In multiple languages, an empty string is the default value of a String, and this is an invalid ID value. However, the default value for Long is 0, which is a valid KEY value. And it's difficult to guarantee that the Zeebe client implementation for each language will override the default KEY value to -1. So it's easier to check the ID first, and the KEY second.

File: clients/go/pkg/commands/evaluate_decision.go (hunk 1 above)
Comment (koevskinikola) [reply clients/go/pkg/commands/evaluate_decision.go:25]: I'm not really sure. I used the `create_instance.go` file as an example of what I needed to do. There is a `Dispatch*Command` [there as well](https://github.com/camunda/zeebe/blob/e9b203656e6787d434aff132a8a19b401e581a08/clients/go/pkg/commands/create_instance.go#L26-L32), even though it doesn't seem necessary. It's the same for the [Complete Job command](https://github.com/camunda/zeebe/blob/f831ee5cb54bc14496e19018e4dad83b64d3c9fc/clients/go/pkg/commands/complete_job.go#L24-L26).

Comment (remcowesterhoud) [reply clients/go/pkg/commands/evaluate_decision.go:25]: Okay, let's just stick to what we're doing in other cases then 🤷

**Comment by koevskinikola** [comment]
bors merge

**Pull request merged**